### PR TITLE
fix(chat): fix completion url space validation (Issue #2382)

### PR DESCRIPTION
**Pull request by Derikyan**
**Description:**

[Custom, Code apps]: Spaces at beginning and end of the url should be ignored. Code app fail to deploy if url has spaces

Issues:

- Issue #2382 

**Checklist:**

- [x] the pull request name complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] the pull request name starts with `fix(<scope>):`, `feat(<scope>):`, `feature(<scope>):`, `chore(<scope>):`, `hotfix(<scope>):` or `e2e(<scope>):`. If contains breaking changes then the pull request name must start with `fix(<scope>)!:`, `feat(<scope>)!:`, `feature(<scope>)!:`, `chore(<scope>)!:`, `hotfix(<scope>)!:` or `e2e(<scope>)!:` where `<scope>` is name of affected project: `chat`, `chat-e2e`, `overlay`, `shared`, `sandbox-overlay`, etc.
- [x] the pull request name ends with `(Issue #<TICKET_ID>)` (comma-separated list of issues)
- [x] I confirm that do not share any confidential information like API keys or any other secrets and private URLs

### CODE REVIEW

**Comment by Derikyan** [comment]
/deploy-review
>Environment: [review-environment](https://chat-ai-dial-chat-pr-3716.nightly-test.deltixhub.io) | [pipeline](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/pipelines/601073)
>Overlay E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1802360)
>Chat E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1802359)

File: apps/chat/src/components/AppsEditor/Settings/ApplicationView/index.tsx [hunk 1]
```diff
@@ -271,6 +268,54 @@ export const ApplicationView: React.FC<Props> = ({ oldApplication }) => {
           id="completionUrl"
           error={errors.completionUrl?.message}
           data-qa="completion-url"
+          onKeyDown={(e) => {
+            if (e.key === ' ') e.preventDefault();
+          }}
+          onBeforeInput={(e: React.FormEvent<HTMLInputElement>) => {
```
Comment (denys-kolomiitsev) [inline apps/chat/src/components/AppsEditor/Settings/ApplicationView/index.tsx]: In my opinion it would be better to extract all inline handlers and maybe wrap in useCallback hook if needed.

**Comment by Derikyan** [comment]
/deploy-review
>Environment: [review-environment](https://chat-ai-dial-chat-pr-3716.nightly-test.deltixhub.io) | [pipeline](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/pipelines/601971)
>Overlay E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1807380)
>Chat E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1807379)

**Review by IlyaBondar — COMMENTED** [review]
## Copiltot answer:

Based on the diff provided, the changes aim to prevent users from entering or pasting spaces into the input field for the `completionUrl`. While the functionality seems to address the requirement, there are areas where the code could be simplified for better readability and maintainability. Below is a review of the changes:

### Review of Existing Changes

1. **`handleUrlBeforeInput`**
   - Prevents the user from typing a space or any input containing a space (`' '` or `'. '`).
   - While functional, it overlaps with `handleUrlKeyDown` and can potentially be removed to simplify the logic.

2. **`handleUrlKeyDown`**
   - Prevents the spacebar key from inserting spaces.
   - This is straightforward and serves its purpose well.

3. **`handleUrlInput`**
   - Cleans any spaces and consecutive dots from the input value. It also updates the cursor position to ensure a smooth user experience.
   - This is functional but contains redundant logic, such as the explicit `setter` call to modify the input value, which can be simplified using React state or a more declarative approach.

4. **`handleUrlPaste`**
   - Prevents pasting spaces by cleaning the pasted content before inserting it into the input field.
   - This is functional but can be streamlined by consolidating space-cleaning logic with `handleUrlInput`.

### Suggestions for Optimization

The following optimizations can reduce complexity while maintaining functionality:

#### Consolidate Logic
You can merge the handling of spaces and invalid characters into a single function that is called on `onInput`, `onPaste`, and `onKeyDown`. For example:

```tsx
const handleUrlChange = (e: React.FormEvent<HTMLInputElement>) => {
  const input = e.currentTarget;
  const cleaned = input.value.replace(/\s+/g, '').replace(/\.{2,}/g, '.');
  
  if (input.value !== cleaned) {
    input.value = cleaned;
    input.dispatchEvent(new Event('input', { bubbles: true }));
  }
};

const handlePreventSpace = (e: React.KeyboardEvent<HTMLInputElement> | React.ClipboardEvent<HTMLInputElement>) => {
  if (e.type === 'keydown' && e.key === ' ') {
    e.preventDefault();
  } else if (e.type === 'paste') {
    e.preventDefault();
    const pasted = (e as React.ClipboardEvent<HTMLInputElement>).clipboardData.getData('text').replace(/\s+/g, '');
    const input = e.currentTarget;
    input.setRangeText(pasted, input.selectionStart ?? 0, input.selectionEnd ?? 0, 'end');
    input.dispatchEvent(new Event('input', { bubbles: true }));
  }
};
```

#### Simplify Event Bindings
You can remove redundant event handlers by combining `onPaste` and `onKeyDown` with a single handler for `onInput`:

```tsx
<input
  id="completionUrl"
  error={errors.completionUrl?.message}
  data-qa="completion-url"
  onKeyDown={handlePreventSpace}
  onPaste={handlePreventSpace}
  onInput={handleUrlChange}
/>
```

### Key Improvements
1. **Readability**: Consolidating logic makes the code easier to understand and reduces duplication.
2. **Performance**: Fewer event listeners reduce potential performance overhead.
3. **Maintainability**: Any future changes to the cleaning logic are centralized.

### Final Code Example
Here’s how the optimized code might look:

```tsx
const handleInputChange = (e: React.FormEvent<HTMLInputElement>) => {
  const input = e.currentTarget;
  const cleaned = input.value.replace(/\s+/g, '').replace(/\.{2,}/g, '.');
  
  if (input.value !== cleaned) {
    input.value = cleaned;
    input.dispatchEvent(new Event('input', { bubbles: true }));
  }
};

const handlePreventSpace = (e: React.KeyboardEvent<HTMLInputElement> | React.ClipboardEvent<HTMLInputElement>) => {
  if (e.type === 'keydown' && e.key === ' ') {
    e.preventDefault();
  } else if (e.type === 'paste') {
    e.preventDefault();
    const pasted = (e as React.ClipboardEvent<HTMLInputElement>).clipboardData.getData('text').replace(/\s+/g, '');
    const input = e.currentTarget;
    input.setRangeText(pasted, input.selectionStart ?? 0, input.selectionEnd ?? 0, 'end');
    input.dispatchEvent(new Event('input', { bubbles: true }));
  }
};

// JSX
<input
  id="completionUrl"
  error={errors.completionUrl?.message}
  data-qa="completion-url"
  onKeyDown={handlePreventSpace}
  onPaste={handlePreventSpace}
  onInput={handleInputChange}
/>
```

This optimized version significantly reduces complexity while maintaining the intended functionality.

**Comment by Derikyan** [comment]
I agree with all the suggested optimizations except for removing onBeforeInput, because without it, handling edge cases like input starting with `.` or typing `. ` may not work correctly.

**Comment by Derikyan** [comment]
/deploy-review
>Environment: [review-environment](https://chat-ai-dial-chat-pr-3716.nightly-test.deltixhub.io) | [pipeline](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/pipelines/602262)
>Chat E2E tests status: [failed](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1808910)
>Overlay E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1808911)

**Comment by Derikyan** [comment]
/deploy-review
>Environment: [review-environment](https://chat-ai-dial-chat-pr-3716.nightly-test.deltixhub.io) | [pipeline](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/pipelines/602335)
>Overlay E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1809416)
>Chat E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1809415)

File: apps/chat/src/components/AppsEditor/Settings/ApplicationView/index.tsx [hunk 2]
```diff
@@ -180,6 +177,49 @@ export const ApplicationView: React.FC<Props> = ({ oldApplication }) => {
...
+      if (keyEvent.key === ' ') {
+        e.preventDefault();
+      }
+    } else if (e.type === 'paste') {
+      e.preventDefault();
+      const pasted = (e as React.ClipboardEvent<HTMLInputElement>).clipboardData
+        .getData('text')
+        .replace(/\s+/g, '');
+      const input = e.currentTarget;
+      input.setRangeText(
```
Comment (IlyaBondar) [inline apps/chat/src/components/AppsEditor/Settings/ApplicationView/index.tsx]: all these methods should be wrapped into `useCallback` to not create each time

and maybe we can move them into custom hook like `const { onBeforeInput, onKeyDown, onPaste, onInput } = usePreventSpaceHandlers()` (where onKeyDown === onPaste)

**Comment by Derikyan** [comment]
/deploy-review
>Environment: [review-environment](https://chat-ai-dial-chat-pr-3716.nightly-test.deltixhub.io) | [pipeline](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/pipelines/603380)
>Overlay E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1815232)
>Chat E2E tests status: [success](https://gitlab.deltixhub.com/Clients/deltix/nightly-test/dial/review/-/jobs/1815231)

File: apps/chat/src/components/AppsEditor/Settings/ApplicationView/index.tsx (hunk 2 above)
Comment (Derikyan) [reply apps/chat/src/components/AppsEditor/Settings/ApplicationView/index.tsx]: thanks, this is a good suggestion for optimization

**Review by IlyaBondar — APPROVED** [review]
